### ISSUE TITLE
Are classes not supported or have I done something wrong?

### ISSUE THREAD

**Issue by MattBarkway** [mathsman5133/coc.py#47]
I wanted to make use of your API wrapper, but instead of the flat functions and decorator setup you have in your examples, I wanted to have mine setup as a class inheriting from EventsClient.
However it doesn't seem very happy, and from within the coc.py library I'm getting lots of 

> TypeError: reset_event_cache() missing 1 required positional argument: 'cache_name'

and other 'missing 1 required positional argument' type errors.

Here's a simple version of my code that  demonstrates the issue

`

    import asyncio
    from coc import EventsClient
    
    
    class ClashConnection(EventsClient):
        def __init__(self, username, password, clan_tag):
            super().__init__()
            self.username = username
            self._password = password
            self.clan_tag = clan_tag
            self.function_mappings = {
                'on_clan_update': self.clan_updated,
            }
    
        async def start(self):
            await self.login(self.username, self._password)
            self.add_clan_update(self.clan_tag, retry_interval=30)  # <-- this bit is the first place it falls over
            self.add_events(function_dicts=self.function_mappings)
            self.start_updates()
    
        @staticmethod
        async def clan_updated():
            print('clan was updated')
    
    
    if __name__ == "__main__":
        import argparse
    
        parser = argparse.ArgumentParser()
        parser.add_argument('username')
        parser.add_argument('password')
        parser.add_argument('clan_tag')
        args = parser.parse_args()
    
        cc = ClashConnection(args.username, args.password, args.clan_tag)
        loop = asyncio.get_event_loop()
        loop.run_until_complete(cc.start())

`

Is this something that you don't support, or am I just missing something obvious?

**Comment by mathsman5133** [comment]
Hi! Thanks for the issue.

Yes, subclassing is supported, but the preferred way of doing it is passing your subclass to `coc.login`, for example:

```py
class ClashConnection(EventsClient):
    def __init__(self, clan_tag):
        super().__init__()
        self.clan_tag = clan_tag
        self.function_mappings = {
            'on_clan_update': self.clan_updated,
        }

    async def login(self, email, password):
        super().login(email, password)
        self.add_clan_update(self.clan_tag, retry_interval=30)  # <-- this bit is the first place it falls over
        self.add_events(function_dicts=self.function_mappings)
        self.start_updates()

    @staticmethod
    async def clan_updated(old_clan, new_clan):
        print('clan was updated')

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument('username')
    parser.add_argument('password')
    parser.add_argument('clan_tag')
    args = parser.parse_args()

    cc = coc.login(args.username, args.password, client=ClashConnection, clan_tag=args.clan_tag)
    cc.run_forever()
```

I'll run through a couple things of note that are different to the example you posted:

1. This now overrides `ClashConnection.login`, which is called in `coc.login` when you login. This means that anything you added in there is automatically done when logging in, as well as `coc.login` calling anything else it needs to setup. The reason for your issue is that [coc.login actually sets up a cache instance before logging in](https://github.com/mathsman5133/coc.py/blob/35cd26d2a34ab877e12cbf861846a6eeab75cfe0/coc/login.py#L55), which meant that as you didn't do that, the `reset_event_cache` method was a member of an uninitiated class, meaning it was looking for a `self` parameter.

2. As per [the docs](https://cocpy.readthedocs.io/en/latest/api.html#coc.on_clan_update), the `on_clan_update` event requires 2 arguments: an old/stale clan, and a new clan. This would have failed if you didn't have that.

3. You need a `cc.run_forever()` line to run the loop forever, otherwise it gets to the end and just exits, rather than running the loop forever. This should be the last function that ever gets called.


tl;dr You didn't call [this](https://github.com/mathsman5133/coc.py/blob/35cd26d2a34ab877e12cbf861846a6eeab75cfe0/coc/login.py#L55) that created the cache instance 👍

**Comment by MattBarkway** [comment]
Thank you for the quick response, that all works now cheers 👍
